include/mata/nfa.hh

```diff
@@ -271,6 +315,44 @@ public:
         return post[q];
     };
 
+    // TODO: why do we have the code of all these methods in the header file? Should we move it out?
+    const std::vector<State> defragment(Util::NumberPredicate<State> & is_staying) {
+
+        //first, indexes of post are filtered (places of to be removed states are taken by states on their right)
+        Util::filter_indexes(post, [is_staying](State i) { return is_staying[i]; });
+
+        //get renaming of current states to new numbers:
+        std::vector<State> renaming(this->find_max_state()+1);
+        size_t i = 0;
+        for (State q:is_staying.get_elements()) {
+            if (q >= renaming.size())
+                break;
+            renaming[q] = i;
+            i++;
+        }
+
+        //this iterates through every post and ever, filters and renames states,
+        //and finally removes moves that became empty from the post.
+        for (State q=0;q<post.size();++q) {
+            //should we have a function Post::transform(Lambda) for this?
+            Post & post = get_mutable_post(q);
+            for (auto move = post.begin(); move < post.end();move++) {
```

Comment:
`++move` is probably much faster in this case.

Reply:
ok